DatadogInternal/Sources/NetworkInstrumentation/NetworkInstrumentationFeature.swift

```diff
@@ -49,80 +52,72 @@ internal final class NetworkInstrumentationFeature: DatadogFeature {
     ///      use the instance of the delegate to get the type.
-    internal func bindIfNeeded(configuration: URLSessionInstrumentation.Configuration) throws {
+    internal func bind(configuration: URLSessionInstrumentation.Configuration) throws {
         let configuredFirstPartyHosts = FirstPartyHosts(firstPartyHosts: configuration.firstPartyHostsTracing) ?? .init()
 
-        try URLSessionTaskDelegateSwizzler.bindIfNeeded(
-            delegateClass: configuration.delegateClass,
-            interceptDidFinishCollecting: { [weak self] session, task, metrics in
-                self?.queue.async { [weak self, weak session] in
-                    self?._task(task, didFinishCollecting: metrics)
-                    session?.delegate?.interceptor?.task(task, didFinishCollecting: metrics)
+        let identifier = ObjectIdentifier(configuration.delegateClass)
 
-                    // iOS 16 and above, didCompleteWithError is not called hence we use task state to detect task completion
-                    // while prior to iOS 15, task state doesn't change to completed hence we use didCompleteWithError to detect task completion
-                    if #available(iOS 15, tvOS 15, *) {
-                        self?._task(task, didCompleteWithError: task.error)
-                        session?.delegate?.interceptor?.task(task, didCompleteWithError: task.error)
-                    }
+        if let swizzler = swizzlers[identifier] {
+            DD.logger.warn(
+                """
+                The delegate class \(configuration.delegateClass) is already instrumented.
+                The previous instrumentation will be disabled in favor of the new one.
+                """
+            )
+
+            swizzler.unswizzle()
+        }
+
+        let swizzler = NetworkInstrumentationSwizzler()
+        swizzlers[identifier] = swizzler
+
+        try swizzler.swizzle(
+            interceptResume: { [weak self] task in
+                // intercept task if delegate match
+                guard let self = self, task.dd.delegate?.isKind(of: configuration.delegateClass) == true else {
+                    return
                 }
-            }, interceptDidCompleteWithError: { [weak self] session, task, error in
-                self?.queue.async { [weak self, weak session] in
-                    // prior to iOS 15, task state doesn't change to completed
-                    // hence we use didCompleteWithError to detect task completion
-                    self?._task(task, didCompleteWithError: task.error)
-                    session?.delegate?.interceptor?.task(task, didCompleteWithError: task.error)
+
+                if let currentRequest = task.currentRequest {
+                    let request = self.intercept(request: currentRequest, additionalFirstPartyHosts: configuredFirstPartyHosts)
+                    task.dd.override(currentRequest: request)
                 }
```

Comment:
question/ This condition is quite puzzling. In what circumstances we expect `task` to not have `currentRequest`? Let's explain it in the code comment if appropriate.

Reply:
That's a good question, I don't know what could nillify the `originalRequest ` or `currentRequest` 🤔  the doc doesn't say..